### PR TITLE
[Merged by Bors] - feat(CategoryTheory): characterise the covering sieves for the extensive topology

### PR DESCRIPTION
**Pull request by dagurtomas**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib.lean [hunk 1]
```diff
@@ -1614,6 +1614,8 @@ import Mathlib.CategoryTheory.Sites.Coherent.CoherentTopology
 import Mathlib.CategoryTheory.Sites.Coherent.Comparison
 import Mathlib.CategoryTheory.Sites.Coherent.Equivalence
 import Mathlib.CategoryTheory.Sites.Coherent.ExtensiveSheaves
+import Mathlib.CategoryTheory.Sites.Coherent.ExtensiveTopology
+import Mathlib.CategoryTheory.Sites.Coherent.LocallySurjective
```
Comment (dagurtomas) [inline Mathlib.lean:1618]:
```suggestion
```

File: Mathlib.lean [hunk 2]
```diff
@@ -1800,6 +1802,7 @@ import Mathlib.Computability.TMToPartrec
 import Mathlib.Computability.TuringMachine
 import Mathlib.Condensed.Basic
 import Mathlib.Condensed.Discrete
+import Mathlib.Condensed.Epi
```
Comment (dagurtomas) [inline Mathlib.lean:1805]:
```suggestion
```

File: Mathlib/CategoryTheory/Sites/LocallySurjective.lean [hunk 3]
```diff
@@ -182,6 +182,11 @@ lemma comp_isLocallySurjective_iff
     IsLocallySurjective J (f₁ ≫ f₂) ↔ IsLocallySurjective J f₂ :=
   isLocallySurjective_iff_of_fac J rfl
 
+variable {J} in
+lemma isLocallySurjectiveMono {K : GrothendieckTopology C} (hJK : J ≤ K) {F G : Cᵒᵖ ⥤ A}
+    (f : F ⟶ G) (h : IsLocallySurjective J f) : IsLocallySurjective K f where
+  imageSieve_mem s := by apply hJK; exact h.1 _
+
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/Sites/LocallySurjective.lean:189]:
```suggestion
```

File: Mathlib/CategoryTheory/Sites/Coherent/ExtensiveSheaves.lean [hunk 4]
```diff
@@ -131,4 +131,8 @@ theorem Presheaf.isSheaf_iff_preservesFiniteProducts {D : Type*} [Category D]
     rw [Presieve.isSheaf_iff_preservesFiniteProducts]
     exact ⟨inferInstance⟩
 
+noncomputable instance {D : Type*} [Category D] [FinitaryExtensive C]
```
Comment (riccardobrasca) [inline Mathlib/CategoryTheory/Sites/Coherent/ExtensiveSheaves.lean:134]: Can you add {D : Type*}.. in the beginning of the file (it is also used in the previous thm)? Moreover I think there are useless universes variables.

File: Mathlib/CategoryTheory/Limits/Shapes/Products.lean [hunk 5]
```diff
@@ -291,6 +291,38 @@ def Cofan.isColimitOfIsIsoSigmaDesc {f : β → C} [HasCoproduct f] (c : Cofan f
   IsColimit.ofIsoColimit (colimit.isColimit (Discrete.functor f))
     (Cofan.ext (@asIso _ _ _ _ _ hc) (fun _ => colimit.ι_desc _ _))
 
+lemma Cofan.isColimit_iff_isIso_sigmaDesc
+    {β : Type*} {f : β → C} [HasCoproduct f] (c : Cofan f) :
+    IsIso (Sigma.desc c.inj) ↔ Nonempty (IsColimit c) := by
+  refine ⟨fun h ↦ ⟨isColimitOfIsIsoSigmaDesc c⟩, fun ⟨hc⟩ ↦ ?_⟩
```
Comment (riccardobrasca) [inline Mathlib/CategoryTheory/Limits/Shapes/Products.lean:297]:
```suggestion
lemma Cofan.isColimit_iff_isIso_sigmaDesc {f : β → C} [HasCoproduct f] (c : Cofan f) :
    IsIso (Sigma.desc c.inj) ↔ Nonempty (IsColimit c) := by
  refine ⟨fun h ↦ ⟨isColimitOfIsIsoSigmaDesc c⟩, fun ⟨hc⟩ ↦ ?_⟩
```

File: Mathlib/CategoryTheory/Limits/Shapes/Products.lean [hunk 6]
```diff
@@ -291,6 +291,38 @@ def Cofan.isColimitOfIsIsoSigmaDesc {f : β → C} [HasCoproduct f] (c : Cofan f
...
+    IsColimit.desc_self, Category.comp_id]
+  rfl
+
+/-- A coproduct of coproducts is a coproduct -/
+def Cofan.isColimitTrans {α : Type*} {X : α → C} (c : Cofan X) (hc : IsColimit c)
+    {β : α → Type*} {Y : (a : α) → β a → C} (π : (a : α) → (b : β a) → Y a b ⟶ X a)
+      (hs : ∀ a, IsColimit (Cofan.mk (X a) (π a))) :
+        IsColimit (Cofan.mk (f := fun ⟨a,b⟩ => Y a b) c.pt
+          (fun (⟨a, b⟩ : Σ a, _) ↦ π a b ≫ c.inj a)) := by
+  refine mkCofanColimit _ ?_ ?_ ?_
```
Comment (riccardobrasca) [inline Mathlib/CategoryTheory/Limits/Shapes/Products.lean:315]:
```suggestion
def Cofan.isColimitTrans {X : α → C} (c : Cofan X) (hc : IsColimit c)
    {β : α → Type*} {Y : (a : α) → β a → C} (π : (a : α) → (b : β a) → Y a b ⟶ X a)
      (hs : ∀ a, IsColimit (Cofan.mk (X a) (π a))) :
        IsColimit (Cofan.mk (f := fun ⟨a,b⟩ => Y a b) c.pt
          (fun (⟨a, b⟩ : Σ a, _) ↦ π a b ≫ c.inj a)) := by
  refine mkCofanColimit _ ?_ ?_ ?_
```

**Comment by riccardobrasca** [comment]
Thanks!

bors d+

**Comment by dagurtomas** [comment]
bors merge

**Pull request closed without merging**